acme/domain.go

```diff
@@ -0,0 +1,247 @@
+package acme
+
+import (
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"net"
+	"strings"
+
+	"github.com/miekg/dns"
+	"golang.org/x/net/publicsuffix"
+)
+
+var recursiveNameserver = "google-public-dns-a.google.com:53"
+
+// dnsQuery sends a DNS query to the given nameserver.
+// The nameserver should include a port, to facilitate testing where we talk to a mock dns server.
+func DnsQuery(fqdn string, rtype uint16, nameserver string, recursive bool) (in *dns.Msg, err error) {
+	m := new(dns.Msg)
+	m.SetQuestion(fqdn, rtype)
+	m.SetEdns0(4096, false)
+
+	if !recursive {
+		m.RecursionDesired = false
+	}
+
+	in, err = dns.Exchange(m, nameserver)
+	if err == dns.ErrTruncated {
+		tcp := &dns.Client{Net: "tcp"}
+		in, _, err = tcp.Exchange(m, nameserver)
+	}
+
+	return
+}
+
+// toFqdn converts the name into a fqdn appending a trailing dot.
+func ToFqdn(name string) string {
+	n := len(name)
+	if n == 0 || name[n-1] == '.' {
+		return name
+	}
+	return name + "."
+}
+
+// unFqdn converts the fqdn into a name removing the trailing dot.
+func UnFqdn(name string) string {
+	n := len(name)
+	if n != 0 && name[n-1] == '.' {
+		return name[:n-1]
+	}
+	return name
+}
+
+
+type Domain struct {
+	Domain            string
+	authoritativeZone string
+	nameServers       []string
+}
+
+func NewDomain(domain string) *Domain {
+	return &Domain{Domain: domain}
+}
+
+// DNS01Record returns a DNS record which will fulfill the `dns-01` challenge
+func (d *Domain) GetDNS01Record(keyAuth string) (fqdn string, value string, ttl int) {
+	keyAuthShaBytes := sha256.Sum256([]byte(keyAuth))
+	// base64URL encoding without padding
+	keyAuthSha := base64.URLEncoding.EncodeToString(keyAuthShaBytes[:sha256.Size])
+	value = strings.TrimRight(keyAuthSha, "=")
+	ttl = 120
+	fqdn = fmt.Sprintf("_acme-challenge.%s.", d.Domain)
+	return
+}
+
+func (d *Domain) GetFqdn() string {
+	return ToFqdn(d.Domain)
+}
+
+// LookupNameservers returns the authoritative nameservers for the given fqdn.
+func (d *Domain) GetNameservers() ([]string, error) {
+	var authoritativeNss []string
+
+	zone, err := d.GetAuthoritativeZone()
+	if err != nil {
+		return nil, err
+	}
+
+	r, err := DnsQuery(zone, dns.TypeNS, recursiveNameserver, true)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, rr := range r.Answer {
+		if ns, ok := rr.(*dns.NS); ok {
+			authoritativeNss = append(authoritativeNss, strings.ToLower(ns.Ns))
```

Comment:
Also, can we add the ":53" here? So the resulting list of nameservers have a proper port.